### PR TITLE
compound: replace boost ranges with std ranges

### PR DESCRIPTION
**Pull request by avikivity**
Continue standardization on std::ranges. Since compound contains a custom
iterator, we first have to upgrade it to C++20 iterator concepts.

Cleanup / minor refactoring, so no backport.

### CODE REVIEW

File: compound.hh [hunk 1]
```diff
@@ -187,8 +186,7 @@ public:
             ++(*this);
             return i;
         }
-        const value_type& operator*() const { return _current; }
-        const value_type* operator->() const { return &_current; }
```
Comment (nyh) [inline compound.hh:191]: ּRemoving this operator forced you to make dozens of ugly changes to the code below. It's sad that it's necessary... The discussion about "stable pointers" doesn't apply to any of the uses of this operator-> that we had in the code, as far as I can tell.

**Review by nyh — APPROVED** [review]

File: compound.hh [hunk 2]
```diff
@@ -187,8 +186,7 @@ public:
...
+        value_type operator*() const { return _current; }
```
[the remark this answers is not in the document]
Comment (avikivity) [reply compound.hh:188]: Yes (and would be eliminated by the optimizer)

File: compound.hh (hunk 1 above)
Comment (avikivity) [reply compound.hh:191]: It's not good to promise something you don't keep.

**Comment by avikivity** [comment]
@tgrabiec review ping

**Pull request closed without merging**